### ISSUE TITLE
AttributeError: Can't get attribute 'Samap_Iter' on <module 'samap.mapping' from [path to samap package]

### ISSUE THREAD

**Issue by philoel** [atarashansky/SAMap#37]
Hey! Fun package, excited to work with it. 

A colleague ran SAMap for me on two species so we could see what I might expect to see with my animals. It looks like it worked, and there are some things that are curious that I want to look at more deeply that simply with a Sankey plot. 

To help me with this, the colleague passed me a pickle with the computed SAMap outputs. I'm attaching two jupyter notebooks: SAMap_1_edited is what he wrote to produce the pickle file. SAMap_2_edited is what he told me I should run, in order to load the pickle file and then from there examine what genes are driving matches between species, etc.. 

Sadly, the pickle won't load. In case the jupyter notebook doesn't show the error when I send it to you, I'm also screenshotting it. For reference, I am not a python user (I can run code, but only if the code is provided, and I have little troubleshooting ability). 

![Screenshot 2021-05-31 at 14 54 05](https://user-images.githubusercontent.com/46672796/120196387-16ab5000-c220-11eb-8fd9-e73dc2b19dcb.png)

[Archive.zip](https://github.com/atarashansky/SAMap/files/6570343/Archive.zip)

The colleague produced the pickled object with samap version 0.1.6, and has confirmed he can still open the pickle just fine. I wiped my conda environment and rebuilt it, this time downgrading my samap to 0.1.6 (from 0.1.7) with no improvement.

Any idea of what I can do to open this pickle? I'd love to dig around the data and figure out what causes some cell types to match to each other (one would not have expected the match).

Thanks,
Phil

**Comment by atarashansky** [comment]
Unfortunately, that’s the messy thing about pickle. In 0.1.7 I refactored the Samap_Iter class name. It should be working on 0.1.6, though.

On 0.1.6 can you try importing:
‘from samap.mapping import Samap_Iter’?

**Comment by philoel** [comment]
Alright, from samap version 0.1.6 I opened a cell above the original failing cell, and entered in the text you provided. I think it couldn't find "Samap_Iter" (see screenshot 1), but I looked in the target .py file and noted that there was something called _Samap_Iter so I tried importing _Samap_Iter, which appears to have worked. 

However, that didn't allow the cell with `sm = load_samap('./thepickle.pkl')` to run, still (screenshot 2). 

I also tried `from samap.mapping import _Samap_Iter as Samap_Iter` to try to add in the new thing under the name that might be expected, but that didn't work either (screenshot 3). 

As an aside, my colleague directed me to run md5sum on my pickle file, and on his original. The "hashes"(?) are identical which suggest my pickle isn't corrupted, if that helps rule out some issues. 

![ss1_importSamapIter](https://user-images.githubusercontent.com/46672796/120284855-fda4ac80-c2bc-11eb-8aca-a207d220d3a3.png)
![ss2_import_SamapIter](https://user-images.githubusercontent.com/46672796/120284869-009f9d00-c2bd-11eb-9bc3-fd7c8dd0a947.png)
![ss3_import_SamapIter_as__SamapIter](https://user-images.githubusercontent.com/46672796/120284888-04332400-c2bd-11eb-9c8a-cbdfcbdc9f13.png)

**Comment by atarashansky** [comment]
So that means you’re still on the wrong version.

can you check ‘samap.__version__’ ?

**Comment by philoel** [comment]
Hi! I was just in a good troubleshooting session with my collaborator, and we solved it, and just like you said it turns out I didn't manage to change versions, though I thought I did. 

So, it's working now (we can close the issue), but I'll summarize what I did and didn't do, in case it helps someone else who makes my errors. 

First, the reason I was still on version 0.1.7 even though I thought I was on 0.1.6: I didn't know how to install a targeted version, so when I cloned the repo locally, before installing its contents I just nagivated to setup.py and changed the information on line5 manually to say `version="0.1.6"` , which I thought would tell the installer what version to fetch. I didn't understand that this only changes the name of the version displayed in my conda environment (which of course told me I was working with version="0.1.6" thereafter). So, the real solution was to manually toggle to the version="0.1.6" page on github, download the repo, then install that manually. That is how to truly install version="0.1.6" and not simply tell your setup.py to name it the wrong thing. 

There was a subsequent issue: even after I manually (and correctly) installed samap version="0.1.6" , at some point in loading the pickle we had an error that was traceable to the installation version of "numba". I had version = 0.53.1, and my collaborator had version 0.52.0. When I downgraded my numba package, suddenly I could work with and process the samap pickle and access the data. 

So in summary, if a samap version 0.1.7 user is using a pickle with the outputs of SAMap that was produced by a user of an older version, take all possible precautions to replicate their conda environment perfectly, and pay special attention to matching the samap package version and the numba package version. 

What a journey.
